### PR TITLE
Deprecate and replace uses of attrib helpers in attrutils

### LINKED ISSUE #44: Document that attrutils are deprecated; document replacements
This will require first finishing https://github.com/isi-vista/immutablecollections/issues/12 (we can omit type checking support for now)

### PR DESCRIPTION
**Pull request by ConstantineLignos**
Closes #44

### CODE REVIEW

**Comment by ConstantineLignos** [comment]
@rgabbard This does the deprecation, but I don't think I'll have time to replace the uses. Do you have anyone on your projects with time for it? There are about 36 in this repo and they should be straightforward.

**Comment by rgabbard** [comment]
@nicomitchell : Can you handle replacing the instances of these methods in `vistautils` that @ConstantineLignos requests above?

File: vistautils/attrutils.py [hunk 1]
```diff
@@ -1,15 +1,17 @@
...
 
-import immutablecollections
 import vistautils.preconditions
 
 
-# TODO cannot currently be used with additional validators:
-# https://github.com/isi-nlp/isi-flexnlp/issues/188
 def attrib_instance_of(type_: Union[Type, Tuple[Type, ...]], *args, **kwargs):
+    warnings.warn(
+        "Deprecated, use attrib(validator=instance_of(type)", DeprecationWarning
```
[the remark this answers is not in the document]
Comment (ConstantineLignos) [reply vistautils/attrutils.py]: Fixed in the amended commit 3d4e270

File: vistautils/attrutils.py [hunk 2]
```diff
@@ -21,20 +23,28 @@ def attrib_instance_of(type_: Union[Type, Tuple[Type, ...]], *args, **kwargs):
 def attrib_opt_instance_of(
     type_: Union[Type, Tuple[Type, ...]], *args, default=None, **kwargs
 ):
+    warnings.warn(
+        "Deprecated, use attrib(validator=optional(instance_of(<type>)))",
```
[the remark this answers is not in the document]
Comment (ConstantineLignos) [reply vistautils/attrutils.py]: Good catch. Fixed in the amended commit 3d4e270 .

File: vistautils/io_utils.py [hunk 3]
```diff
@@ -564,16 +564,16 @@ def __exit__(self, exc_type, exc_val, exc_tb):
 
 @attrs(slots=True, frozen=True)
 class _FileCharSink(CharSink):
-    _path: Union[Path, str] = attrib_instance_of((Path, str))
+    _path: Union[Path, str] = attrib(validator=validators.instance_of(tuple((Path, str))))
```
Comment (rgabbard) [inline vistautils/io_utils.py]: Don't need `tuple` here - just `(Path, str)` is already a tuple

File: vistautils/key_value.py [hunk 4]
```diff
@@ -306,15 +305,17 @@ def from_path_mapping(id_to_path: Mapping[str, Path]) -> "KeyValueSource[str, st
...
+        return _PathMappingCharKeyValueSource(id_to_path)  # type: ignore
 
     @staticmethod
     def from_doc_id_to_file_map(
         map_file: Union[str, Path, CharSource]
     ) -> "KeyValueSource[str,str]":
         if not isinstance(map_file, CharSource):
             map_file = CharSource.from_file(map_file)
-        return _PathMappingCharKeyValueSource(read_doc_id_to_file_map(map_file))
+        return _PathMappingCharKeyValueSource(  # type: ignore
```
Comment (rgabbard) [inline vistautils/key_value.py]: Are the `type: ignores` in this file still needed?

File: vistautils/parameters.py [hunk 5]
```diff
@@ -56,7 +55,9 @@ class Parameters:
     You can check if a lookup of a parameter would be successful using the `in` operator.
     """
 
-    _data: ImmutableDict[str, Any] = attrib_opt_immutable(ImmutableDict)
+    _data: ImmutableDict[str, Any] = attrib(
+        validator=validators.instance_of(ImmutableDict)
```
Comment (rgabbard) [inline vistautils/parameters.py]: `attrib(converter=immutabledict, default=immutabledict())`

File: vistautils/range.py [hunk 6]
```diff
@@ -1156,7 +1155,9 @@ def __repr__(self):
 # https://github.com/python-attrs/attrs/issues/313
 @attrs(frozen=True, repr=False)
 class ImmutableRangeMap(Generic[K, V], RangeMap[K, V]):
-    rng_to_val: ImmutableDict[Range[K], V] = attrib_immutable(ImmutableDict)
+    rng_to_val: ImmutableDict[Range[K], V] = attrib(
```
Comment (rgabbard) [inline vistautils/range.py]: `attrib(converter=immutabledict)`

File: vistautils/io_utils.py (hunk 3 above)
Comment (nicomitchell) [reply vistautils/io_utils.py]: that one was throwing a mypy error until I added the `tuple`

Comment (rgabbard) [reply vistautils/io_utils.py]: Huh, weird. Ok, leave it as-is.

File: vistautils/key_value.py (hunk 4 above)
Comment (nicomitchell) [reply vistautils/key_value.py]: Yeah, the mypy error gets thrown at these lines rather than inside the function they're calling

**Pull request merged**